File: domain/src/main/java/org/openmedstack/eventstore/IEventStream.java
package org.openmedstack.eventstore;

import java.util.HashMap;
import java.util.List;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public interface IEventStream extends AutoCloseable {
    String BucketId();

    String getStreamId();

    int getStreamRevision();

    int getCommitSequence();

    List<EventMessage> getCommittedEvents();

    HashMap<String, Object> getCommittedHeaders();

    List<EventMessage> getUncommittedEvents();

    HashMap<String, Object> getUncommittedHeaders();

    void add(EventMessage uncommittedEvent);

    CompletableFuture commitChanges(UUID commitId) throws DuplicateCommitException, ConcurrencyException;

    void clearChanges();
}

File: domain/src/main/kotlin/org/openmedstack/eventstore/IEventStream.kt
package org.openmedstack.eventstore

import java.util.*
import java.util.concurrent.CompletableFuture

interface IEventStream : AutoCloseable {
    val bucketId: String?
    val streamId: String
    val streamRevision: Int
    val commitSequence: Int
    val committedEvents: List<EventMessage>
    val committedHeaders: HashMap<String, Any>
    val uncommittedEvents: List<EventMessage>
    val uncommittedHeaders: HashMap<String, Any>
    fun add(uncommittedEvent: EventMessage)
    @Throws(DuplicateCommitException::class, ConcurrencyException::class)
    fun commitChanges(commitId: UUID): CompletableFuture<*>
    fun clearChanges()
}